File: netDxf/Entities/Hatch.Transform.cs
using System;
using System.Collections.Generic;

namespace netDxf.Entities
{
    public partial class Hatch
    {
        /// <summary>Transforms stored boundary geometry and the hatch plane.</summary>
        /// <param name="transformation">Linear transformation, using column vectors.</param>
        /// <param name="translation">Translation in world coordinates.</param>
        /// <remarks>
        /// Line, polyline, conic and spline boundaries in solid fills support
        /// finite affine transformations that preserve a two-dimensional plane. Spline
        /// knots, weights, rationality, periodicity and optional fit metadata retain
        /// their stored meanings; this operation does not evaluate or refit a curve.
        /// Circular boundaries become ellipse edges when required. Conic results whose
        /// stored angles cannot preserve their endpoints within relative tolerance
        /// 1e-10 are rejected. Explicit predefined/custom pattern line families support
        /// affine changes when their stored WCS Point2d origin remains representable.
        /// Nonuniform user-defined, doubled and gradient patterns are rejected.
        /// A successful transform unlinks associative
        /// sources without transforming or removing those source entities. Validation
        /// completes before unlinking or changing stored state.
        /// </remarks>
        public override void TransformBy(Matrix3 transformation, Vector3 translation)
        {
            for (int row = 0; row < 3; row++)
                for (int column = 0; column < 3; column++)
                    RequireAffineFinite(transformation[row, column]);
            RequireAffineFinite(translation);
            RequireAffineFinite(this.Elevation);
            RequireAffineFinite(this.Pattern.Angle);
            RequireAffineFinite(this.Pattern.Scale);
            ValidateAffinePathData(this.BoundaryPaths);

            Matrix3 oldOcs = MathHelper.ArbitraryAxis(this.Normal);
            Vector3 xAxis = transformation * (oldOcs * Vector3.UnitX);
            Vector3 yAxis = transformation * (oldOcs * Vector3.UnitY);
            RequireAffineFinite(xAxis); RequireAffineFinite(yAxis);
            Vector3 xUnit = AffineUnit(xAxis), yUnit = AffineUnit(yAxis);
            Vector3 normal = AffineUnit(Vector3.CrossProduct(xUnit, yUnit));
            Vector3 normalHint = transformation * this.Normal;
            RequireAffineFinite(normalHint);
            // Either sign describes the plane. Keep the historical orientation where
            // its transformed normal has a nonzero component perpendicular to it.
            if (Vector3.DotProduct(normal, normalHint) < 0) normal = -normal;
            Matrix3 newOcs = MathHelper.ArbitraryAxis(normal).Transpose();
            double xLength = xAxis.Modulus(), yLength = yAxis.Modulus();
            RequireAffineFinite(xLength); RequireAffineFinite(yLength);
            bool similarity = Math.Abs(Vector3.DotProduct(xUnit, yUnit)) <= 1e-12
                && Math.Abs(xLength - yLength) <= 1e-12 * Math.Max(xLength, yLength);
            bool explicitPattern = this.HasExplicitAffinePattern();
            if (!similarity && (this.Pattern is HatchGradientPattern
                || this.Pattern.Fill != HatchFillType.SolidFill && (!explicitPattern || this.Pattern.IsDouble)))
                throw new NotSupportedException("Nonuniform HATCH transforms require a solid fill or an explicit, non-doubled predefined/custom pattern.");

            Vector3 position = newOcs * (transformation * (oldOcs * new Vector3(0, 0, this.Elevation)) + translation);
            RequireAffineFinite(position);
            Func<Vector2, bool, Vector2> map = (value, vector) =>
            {
                RequireAffineFinite(value.X); RequireAffineFinite(value.Y);
                Vector3 point = newOcs * (transformation * (oldOcs * new Vector3(value.X, value.Y, vector ? 0 : this.Elevation)) + (vector ? Vector3.Zero : translation));
                RequireAffineFinite(point);
                return new Vector2(point.X, point.Y);
            };
            var paths = new List<HatchBoundaryPath>();
            foreach (HatchBoundaryPath path in this.BoundaryPaths)
            {
                var edges = new List<HatchBoundaryPath.Edge>();
                foreach (HatchBoundaryPath.Edge edge in path.Edges)
                {
                    if (edge is HatchBoundaryPath.Line line)
                        edges.Add(new HatchBoundaryPath.Line { Start = map(line.Start, false), End = map(line.End, false) });
                    else if (edge is HatchBoundaryPath.Spline spline)
                    {
                        edges.Add(TransformAffineSpline(spline, map));
                    }
                    else if (edge is HatchBoundaryPath.Arc arc)
                        edges.Add(TransformAffineArc(arc, map, similarity));
                    else if (edge is HatchBoundaryPath.Ellipse ellipse)
                        edges.Add(TransformAffineConic(ellipse.Center, ellipse.EndMajorAxis, ellipse.MinorRatio, ellipse.StartAngle, ellipse.EndAngle, ellipse.IsCounterclockwise, map));
                    else
                        TransformAffinePolyline((HatchBoundaryPath.Polyline)edge, map, similarity,
                            AffineCross(map(Vector2.UnitX, true), map(Vector2.UnitY, true)) < 0, edges);
                }
                var transformedPath = new HatchBoundaryPath(edges);
                transformedPath.PathType = (path.PathType & ~HatchBoundaryPathTypeFlags.Polyline)
                    | (transformedPath.PathType & HatchBoundaryPathTypeFlags.Polyline);
                paths.Add(transformedPath);
            }
            ValidateAffinePathData(paths);
            var seeds = new List<Vector2>();
            foreach (Vector2 seed in this.seedPoints) seeds.Add(map(seed, false));
            Vector2 direction = this.Pattern.Scale * Vector2.Rotate(Vector2.UnitX, this.Pattern.Angle * MathHelper.DegToRad);
            Vector2 axis = map(direction, true);
            double scale = axis.Modulus(), angle = Vector2.Angle(axis) * MathHelper.RadToDeg;
            RequireAffineFinite(scale); RequireAffineFinite(angle);
            if (scale <= 0) throw new ArgumentException("The HATCH pattern direction collapses under this transformation.");

            bool identity = translation.X == 0 && translation.Y == 0 && translation.Z == 0;
            for (int row = 0; row < 3; row++)
                for (int column = 0; column < 3; column++)
                    identity &= transformation[row, column] == (row == column ? 1 : 0);
            HatchPattern transformedPattern = explicitPattern
                ? this.TransformAffinePattern(transformation, translation, map, scale, angle, identity) : null;
            // Everything above is temporary. Unlink only after validation succeeds.
            if (this.associative) this.UnLinkBoundary();
            if (identity) return;
            if (transformedPattern != null) this.pattern = transformedPattern;
            else { this.Pattern.Scale = scale; this.Pattern.Angle = angle; }
            this.Elevation = position.Z; this.Normal = normal;
            this.BoundaryPaths.Clear(); this.BoundaryPaths.AddRange(paths);
            this.seedPoints.Clear(); foreach (Vector2 seed in seeds) this.seedPoints.Add(seed);
        }

        private static HatchBoundaryPath.Spline TransformAffineSpline(HatchBoundaryPath.Spline spline, Func<Vector2, bool, Vector2> map)
        {
                        HatchSplineData.Validate(spline);
                        var copy = new HatchBoundaryPath.Spline
                        {
                            Degree = spline.Degree, IsRational = spline.IsRational, IsPeriodic = spline.IsPeriodic,
                            Knots = (double[])spline.Knots.Clone(), ControlPoints = (Vector3[])spline.ControlPoints.Clone(),
                            StartTangent = spline.StartTangent, EndTangent = spline.EndTangent
                        };
                        foreach (Vector2 fit in spline.FitPoints) copy.FitPoints.Add(fit);
                        for (int i = 0; i < copy.ControlPoints.Length; i++)
                        {
                            Vector3 value = copy.ControlPoints[i]; Vector2 point = map(new Vector2(value.X, value.Y), false);
                            copy.ControlPoints[i] = new Vector3(point.X, point.Y, value.Z);
                        }
                        for (int i = 0; i < copy.FitPoints.Count; i++) copy.FitPoints[i] = map(copy.FitPoints[i], false);
                        if (copy.StartTangent.HasValue) copy.StartTangent = map(copy.StartTangent.Value, true);
                        if (copy.EndTangent.HasValue) copy.EndTangent = map(copy.EndTangent.Value, true);
                        return copy;
        }

        private static void ValidateAffinePathData(IEnumerable<HatchBoundaryPath> paths)
        {
            foreach (HatchBoundaryPath path in paths)
                foreach (HatchBoundaryPath.Edge edge in path.Edges)
                {
                    if (edge is HatchBoundaryPath.Line line)
                    { RequireAffineFinite(line.Start); RequireAffineFinite(line.End); }
                    else if (edge is HatchBoundaryPath.Spline spline)
                    {
                        HatchSplineData.Validate(spline);
                        foreach (Vector2 fit in spline.FitPoints) RequireAffineFinite(fit);
                        if (spline.StartTangent.HasValue) RequireAffineFinite(spline.StartTangent.Value);
                        if (spline.EndTangent.HasValue) RequireAffineFinite(spline.EndTangent.Value);
                    }
                    else if (edge is HatchBoundaryPath.Polyline polyline)
                    {
                        if (polyline.Vertexes == null) throw new ArgumentException("HATCH polyline vertices must be supplied.");
                        foreach (Vector3 vertex in polyline.Vertexes) RequireAffineFinite(vertex);
                    }
                    else if (edge is HatchBoundaryPath.Arc arc)
                    { RequireAffineFinite(arc.Center); RequireAffineFinite(arc.Radius); RequireAffineFinite(arc.StartAngle); RequireAffineFinite(arc.EndAngle); }
                    else if (edge is HatchBoundaryPath.Ellipse ellipse)
                    { RequireAffineFinite(ellipse.Center); RequireAffineFinite(ellipse.EndMajorAxis); RequireAffineFinite(ellipse.MinorRatio); RequireAffineFinite(ellipse.StartAngle); RequireAffineFinite(ellipse.EndAngle); }
                    else throw new ArgumentException("Unsupported or absent HATCH boundary edge.");
                }
        }

        private static Vector3 AffineUnit(Vector3 value)
        {
            double largest = Math.Max(Math.Abs(value.X), Math.Max(Math.Abs(value.Y), Math.Abs(value.Z)));
            if (largest == 0) throw new ArgumentException("The HATCH plane collapses under this transformation.");
            RequireAffineFinite(largest);
            value /= largest;
            return value / value.Modulus();
        }
        private static void RequireAffineFinite(double value)
        {
            if (double.IsNaN(value) || double.IsInfinity(value)) throw new ArgumentException("HATCH transforms require finite input and representable results.");
        }
        private static void RequireAffineFinite(Vector2 value)
        { RequireAffineFinite(value.X); RequireAffineFinite(value.Y); }
        private static void RequireAffineFinite(Vector3 value)
        { RequireAffineFinite(value.X); RequireAffineFinite(value.Y); RequireAffineFinite(value.Z); }
    }
}

File: javascript/netDxf/Entities/Hatch.Transform.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import {HatchBoundaryPath as H} from './HatchBoundaryPath.js';
import {HatchGradientPattern} from './HatchGradientPattern.js';
import {HatchFillType} from './HatchFillType.js';
import {HatchSplineData} from './HatchSplineData.js';
import {Vector2} from '../Vector2.js';
import {Vector3} from '../Vector3.js';
import {Matrix3} from '../Matrix3.js';
import {MathHelper} from '../MathHelper.js';
import {FixedArray} from '../../runtime/FixedArray.js';
import {DotNetMath as M,MultiplyDouble as mul} from '../../runtime/GeometryRuntime.js';
import {RequireAffineFinite as finite,AffineUnit as unit,AffineCross as cross} from '../../runtime/HatchAffine.js';
import {TransformAffineArc,TransformAffineConic,TransformAffinePolyline} from './Hatch.ConicTransform.js';
import {HasExplicitAffinePattern,TransformAffinePattern} from './Hatch.PatternTransform.js';
import {ArgumentException,NotSupportedException} from '../../runtime/Errors.js';
function validate(paths){
  for(const path of paths)for(const edge of path.Edges){
    if(edge instanceof H.Line){finite(edge.Start);finite(edge.End);}
    else if(edge instanceof H.Spline){HatchSplineData.Validate(edge);for(const p of edge.FitPoints)finite(p);if(edge.StartTangent!==null)finite(edge.StartTangent);if(edge.EndTangent!==null)finite(edge.EndTangent);}
    else if(edge instanceof H.Polyline){if(edge.Vertexes==null)throw new ArgumentException('HATCH polyline vertices must be supplied.');for(const p of edge.Vertexes)finite(p);}
    else if(edge instanceof H.Arc){finite(edge.Center);finite(edge.Radius);finite(edge.StartAngle);finite(edge.EndAngle);}
    else if(edge instanceof H.Ellipse){finite(edge.Center);finite(edge.EndMajorAxis);finite(edge.MinorRatio);finite(edge.StartAngle);finite(edge.EndAngle);}
    else throw new ArgumentException('Unsupported or absent HATCH edge.');
  }
}
function splineCopy(spline,map){
  HatchSplineData.Validate(spline);
  const copy=Object.assign(new H.Spline(),{Degree:spline.Degree,IsRational:spline.IsRational,IsPeriodic:spline.IsPeriodic,Knots:FixedArray(spline.Knots,v=>v),ControlPoints:FixedArray(spline.ControlPoints),StartTangent:spline.StartTangent,EndTangent:spline.EndTangent});
  for(const p of spline.FitPoints)copy.FitPoints.Add(p);
  for(let i=0;i<copy.ControlPoints.length;i++){const v=copy.ControlPoints[i],p=map(new Vector2(v.X,v.Y),false);copy.ControlPoints[i]=new Vector3(p.X,p.Y,v.Z);}
  for(let i=0;i<copy.FitPoints.Count;i++)copy.FitPoints.set_Item(i,map(copy.FitPoints.get_Item(i),false));
  if(copy.StartTangent!==null)copy.StartTangent=map(copy.StartTangent,true);if(copy.EndTangent!==null)copy.EndTangent=map(copy.EndTangent,true);return copy;
}
/** Stage the complete edge/pattern/seed geometry before touching the source model. */
export function TransformHatch(h,transformation,translation){
  for(let r=1;r<=3;r++)for(let c=1;c<=3;c++)finite(transformation[`M${r}${c}`]);
  finite(translation);finite(h.Elevation);finite(h.Pattern.Angle);finite(h.Pattern.Scale);validate(h.BoundaryPaths);
  const oldOcs=MathHelper.ArbitraryAxis(h.Normal),xAxis=Matrix3.Multiply(transformation,Matrix3.Multiply(oldOcs,Vector3.UnitX)),yAxis=Matrix3.Multiply(transformation,Matrix3.Multiply(oldOcs,Vector3.UnitY));
  finite(xAxis);finite(yAxis);const xUnit=unit(xAxis),yUnit=unit(yAxis);let normal=unit(Vector3.CrossProduct(xUnit,yUnit));
  const hint=Matrix3.Multiply(transformation,h.Normal);finite(hint);if(Vector3.DotProduct(normal,hint)<0)normal=Vector3.Negate(normal);
  const newOcs=MathHelper.ArbitraryAxis(normal).Transpose(),xLength=xAxis.Modulus(),yLength=yAxis.Modulus();finite(xLength);finite(yLength);
  const similarity=M.Abs(Vector3.DotProduct(xUnit,yUnit))<=1e-12&&M.Abs(xLength-yLength)<=mul(1e-12,M.Max(xLength,yLength)),explicit=HasExplicitAffinePattern(h);
  if(!similarity&&(h.Pattern instanceof HatchGradientPattern||(h.Pattern.Fill!==HatchFillType.SolidFill&&(!explicit||h.Pattern.IsDouble))))throw new NotSupportedException('Nonuniform transforms require solid fill or explicit nondoubled patterns.');
  const position=Matrix3.Multiply(newOcs,Vector3.Add(Matrix3.Multiply(transformation,Matrix3.Multiply(oldOcs,new Vector3(0,0,h.Elevation))),translation));finite(position);
  const map=(value,vector)=>{
    finite(value.X);finite(value.Y);const point=Matrix3.Multiply(newOcs,Vector3.Add(Matrix3.Multiply(transformation,Matrix3.Multiply(oldOcs,new Vector3(value.X,value.Y,vector?0:h.Elevation))),vector?Vector3.Zero:translation));finite(point);return new Vector2(point.X,point.Y);
  };
  const paths=[];
  for(const path of h.BoundaryPaths){
    const edges=[];
    for(const edge of path.Edges){
      if(edge instanceof H.Line)edges.push(Object.assign(new H.Line(),{Start:map(edge.Start,false),End:map(edge.End,false)}));
      else if(edge instanceof H.Spline)edges.push(splineCopy(edge,map));
      else if(edge instanceof H.Arc)edges.push(TransformAffineArc(edge,map,similarity));
      else if(edge instanceof H.Ellipse)edges.push(TransformAffineConic(edge.Center,edge.EndMajorAxis,edge.MinorRatio,edge.StartAngle,edge.EndAngle,edge.IsCounterclockwise,map));
      else TransformAffinePolyline(edge,map,similarity,cross(map(Vector2.UnitX,true),map(Vector2.UnitY,true))<0,edges);
    }
    const output=H.FromEdges(edges);output.PathType=(path.PathType&~2)|(output.PathType&2);paths.push(output);
  }
  validate(paths);const seeds=Array.from(h.SeedPoints,seed=>map(seed,false));
  const direction=Vector2.Multiply(h.Pattern.Scale,Vector2.Rotate(Vector2.UnitX,mul(h.Pattern.Angle,MathHelper.DegToRad))),axis=map(direction,true),scale=axis.Modulus(),angle=mul(Vector2.Angle(axis),MathHelper.RadToDeg);
  finite(scale);finite(angle);if(scale<=0)throw new ArgumentException('The pattern direction collapses.');
  let identity=translation.X===0&&translation.Y===0&&translation.Z===0;for(let r=1;r<=3;r++)for(let c=1;c<=3;c++)identity=identity&&transformation[`M${r}${c}`]===(r===c?1:0);
  const transformedPattern=explicit?TransformAffinePattern(h,transformation,translation,map,scale,angle,identity):null;
  if(h.Associative)h.UnLinkBoundary();if(identity)return;
  if(transformedPattern!==null)h.Pattern=transformedPattern;else{h.Pattern.Scale=scale;h.Pattern.Angle=angle;}
  h.Elevation=position.Z;h.Normal=normal;h.BoundaryPaths.Clear();h.BoundaryPaths.AddRange(paths);h.SeedPoints.Clear();for(const seed of seeds)h.SeedPoints.Add(seed);
}
